examples/src/main/scala/zio/config/examples/ProgramExample.scala

```diff
@@ -19,77 +20,108 @@ object ProgramExample extends App {
         fail => env.console.putStrLn(s"failed $fail") *> ZIO.succeed(1),
         _ => env.console.putStrLn(s"succeeded") *> ZIO.succeed(0)
       )
-    })
+    }
   }
 }
 
-// spark service
+//// stubs for the real Spark
+
+trait SparkSession {
+  def sql(value: String): DataFrame =
+    new DataFrame {}
+
+  def version: String =
+    "someVersion"
+}
+
+trait DataFrame
+
+////
+
 trait SparkEnv {
   def spark: SparkEnv.Service
 }
 
 object SparkEnv {
-  type SparkSession = String
-
   trait Service {
-    def getSpark: UIO[SparkSession]
+    def sparkEnv: UIO[SparkSession]
   }
 
-  class Live(r: SparkSession) extends Service {
-    def getSpark: UIO[SparkSession] = ZIO.succeed(r)
-  }
+  def make(session: => SparkSession): ZIO[Blocking, Throwable, SparkEnv] =
+    zio.blocking
+      .effectBlocking(session)
+      .map(
+        sparkSession =>
+          new SparkEnv {
+            override def spark: Service =
+              new Service {
+                override def sparkEnv: UIO[SparkSession] =
+                  ZIO.succeed(sparkSession)
+              }
+          }
+      )
+
+  def local(name: String): ZIO[Blocking, Throwable, SparkEnv] =
+    make {
+      // SparkSession.builder().appName(name).enableHiveSupport().getOrCreate()
+      new SparkSession {}
+    }
+
+  def cluster(name: String): ZIO[Blocking, Throwable, SparkEnv] =
+    make {
+      // SparkSession.builder().appName(name).master("local").getOrCreate()
+      new SparkSession {}
+    }
+
 }
 
+////
+
 // The core application
 object Application {
-  val processData: ZIO[SparkEnv with Config[ProgramConfig], Throwable, Unit] =
-    ZIO.accessM(env => {
-      for {
-        conf  <- config[ProgramConfig]
-        spark <- env.spark.getSpark
-        _     <- ZIO.effect(println(s"Executing ${conf.inputPath} and ${conf.outputPath} using ${spark}"))
-      } yield ()
-    })
-
-  val logSparkSession: ZIO[SparkEnv, Throwable, Unit] =
-    ZIO.accessM(
-      r =>
-        for {
-          session <- r.spark.getSpark
-          _ <- ZIO.effect(
-                println(s"Executing something with spark $session without the need of anything else from config")
-              )
-        } yield ()
-    )
-
   val logProgramConfig: ZIO[Config[ProgramConfig], Throwable, Unit] =
     for {
       r <- config[ProgramConfig]
-      _ <- ZIO
-            .effect(
-              println(
-                s"Executing something with programConfig's parameters ${r.inputPath} and ${r.outputPath} without the need of sparkSession"
-              )
+      _ <- ZIO.effect(
+            println(
+              s"Executing something with programConfig's parameters ${r.inputPath} and ${r.outputPath} without the need of sparkSession"
             )
+          )
+    } yield ()
+
+  val runSparkJob: ZIO[SparkEnv with Blocking, Throwable, Unit] =
+    for {
+      session <- ZIO.accessM[SparkEnv](_.spark.sparkEnv)
+      result  <- zio.blocking.effectBlocking(session.sql("SELECT something"))
```

Comment:
session.sql returns a data frame, and it doesn't need to be blocking.
Have a write function and make it blocking

Reply:
Also `zio.blocking.effectBlocking(session.write("somedata"))`